service/history/historybuilder/event_store.go

```diff
@@ -454,43 +453,61 @@ func (b *EventStore) reorderBuffer(
 }
 
 func (b *EventStore) emitInorderedBufferedEvents(bufferedEvents []*historypb.HistoryEvent) {
-	completedActivities := make(map[int64]struct{})
-	completedChildWorkflows := make(map[int64]struct{})
-	var inorderedEventsCount int64
+
+	if b.metricsHandler == nil {
+		return
+	}
+
+	completedActivities := make(map[int64]enumspb.EventType)
+	completedChildWorkflows := make(map[int64]enumspb.EventType)
+	completedNexusOperations := make(map[int64]enumspb.EventType)
+
 	for _, event := range bufferedEvents {
 		switch event.GetEventType() {
+		// Activity.
 		case enumspb.EVENT_TYPE_ACTIVITY_TASK_STARTED:
-			if _, seenCompleted := completedActivities[event.GetEventId()]; seenCompleted {
-				inorderedEventsCount++
+			if completeEventType, seenCompleted := completedActivities[event.GetEventId()]; seenCompleted {
+				metrics.InorderedBufferedEventsCounter.With(b.metricsHandler).Record(1, metrics.OperationTag(completeEventType.String()))
```

Comment:
instead of just metrics, would it be useful to write logs to better understand the situation?

Reply:
I will do this in separate PR because it is much more involved change.